### ISSUE TITLE
Allowance for using Stored Operation IDs for Semantic conventions for GraphQL Server

### ISSUE THREAD

**Issue by magicmark** [open-telemetry/semantic-conventions#1011]
**What are you trying to achieve?**

As an alternative to using the GraphQL query name as the span name, I'd like to use its hash (stored operation ID) instead.

**Context**

We're looking at aligning our GraphQL infra with the following convention:
- https://opentelemetry.io/docs/specs/otel/trace/semantic_conventions/instrumentation/graphql/

However, we use [document IDs (a sha256 of the query)](https://github.com/graphql/graphql-over-http/issues/38
) to uniquely identify queries across our infra. Our org does not allow arbitrary queries at all - every graphql request is known ahead of time, and the query will live somewhere in our codebase.

**Why does this matter in the context of the span name?**

Query names are not unique (different queries can be defined with the same name across different platforms, or even older/multiple versions of the same query) leading to confusion when trying to find the query in source code.

My concern is this:

Imagine a situation where one of our oncall engineers (who is unfamiliar with the GQL stack) is trying to diagnose a slow or failing operation - they might reasonably use the operation name (query name) as the lookup key when trying to find the query in our codebase to debug further. As mentioned above, this is very susceptible to finding the wrong thing. 

**"Just add the operation ID as a tag/attribute"**

While it is possible to do this, there's still a footgun here in that folks have to know to click into the sapn to copy the operation ID instead of the span name.

Generally we try and build a path for folks to guide them to do the right thing by default - and the "right thing by default" here would be putting the operation name as the span name imo.

## The ask

For GraphQL Servers that respect operation IDs only, the span name guidance could allow for using this instead?

## FAQs

**What about batching**

This is actually a different but related point: the first span to our GQL server will be an HTTP request containing multiple queries

https://www.apollographql.com/blog/apollo-client/performance/query-batching

In a case where a server supports batching, I would propose this (where the span name is inside the square brackets)

```
-[ POST /graphql ]
----[ some-long-sha-256-hash-foo ]
----[ some-long-sha-256-hash-bar ]
----[ some-long-sha-256-hash-baz ]
```

**What about Automatic Persisted Queries?**

Apollo support [Automatic Persisted Queries (APQ)](https://www.apollographql.com/docs/apollo-server/performance/apq/) which is a sort of hybrid. The server supports arbitrary queries, with the query name being the correct and only way to identify them. But subsequent requests have their raw query bodies substituted for IDs instead (which the server has stored in a lookup table). My understanding is that in this setup, IDs are not intended to be used as globally unique identifiers - and more similar to a cache key instead. 

tl;dr servers using Apollo APQ should keep the status quo.

**Comment by PascalSenn** [comment]
@magicmark  @SonjaChevre @maraisr There is renewed effort in improving the current open telemetry specification for GraphQL
If you are interested in collaborating comment here: https://github.com/graphql/graphql-wg/discussions/1691

We can then organize a few sessions and submit a proposal
